### ISSUE TITLE
is PYSCF_MAX_MEMORY the memory per core or for the entire system?

### ISSUE THREAD

**Issue by HehnLukas** [pyscf/pyscf#1041]
I've noticed for some parallel computations that when I set the maximum memory, which I thought was for the entire system, close to the total available memory, that my calculations are no longer consistent. By accident I found that reducing the max memory causes them to give consistent results. 
This has made me assume that it sets the memory per core, and that somehow giving it too much memory causes some issues. Can someone confirm if this is the case? I have not found the relevant section of the code.

**Comment by sunqm** [comment]
max_memory is a global setting. Python code uses this parameter for most big array allocations. Certain python calls have significant overhead in memory footprint (like lib.einsum) . Some were considered but there left some parts of the code with bad estimation of memory usage.

C code has extra memory footprint but nothing to do with `max_memory`. Memory allocated in C is not restricted to max_memory. Most C code uses very small amount of local memory per thread. It sometimes leads to more memory usage than `max_memory`.

What's the problem size of your system? NEVPT2 C code roughly needs `n_cas^4` words per thread plus `100 * n_cas^4` words global buffer.

**Comment by HehnLukas** [comment]
Yeah, so I am very confused. I decided to run more tests on a water molecule, as that way the calculations are quicker. It seems if the max memory is set too high, the calculations will no longer be consistent, and for smaller molecules, too high is still at a quite low memory.

My initial tests were with a system consisting of one Fe ion and 6 water molecules, using the def2-TZVP basis set. here, I ran the calculations on 40 cores, and with 150GB of memory allocated to the cluster. When I set PYSCF_MAX_MEMORY to 112000, the resulting casscf orbitals are somewhat different each time, leading to different Nevpt2 energies. However, when I set PYSCF_MAX_MEMORY to 11200, the calculations are identical each time, and the Nev energies stay the same. Even dropping the PYSCF_MAX_MEMORY to 0 results in the same results as 11200. 

With water (here i use 5 cores), I get the same problem, but the threshold is much lower. For example, setting the max_memory to just 500 results in:

Cas1.out:Nevpt2 Energy = -0.248842607769583
Cas2.out:Nevpt2 Energy = -0.247865009848661
Cas3.out:Nevpt2 Energy = -0.248110570899806
Cas4.out:Nevpt2 Energy = -0.246335547520711
Cas5.out:Nevpt2 Energy = -0.248938029269932
Cas6.out:Nevpt2 Energy = -0.248799774140834
Cas7.out:Nevpt2 Energy = -0.248860950140335
Cas8.out:Nevpt2 Energy = -0.248568177855166
Cas9.out:Nevpt2 Energy = -0.247258588547448

whereas using max_mem = 100 (and = 0) I get:

Cas1.out:Nevpt2 Energy = -0.248663823367456
Cas2.out:Nevpt2 Energy = -0.248663823367456
Cas3.out:Nevpt2 Energy = -0.248663823367456
Cas4.out:Nevpt2 Energy = -0.248663823367456
Cas5.out:Nevpt2 Energy = -0.248663823367456
Cas6.out:Nevpt2 Energy = -0.248663823367456
Cas7.out:Nevpt2 Energy = -0.248663823367456
Cas8.out:Nevpt2 Energy = -0.248663823367456
Cas9.out:Nevpt2 Energy = -0.248663823367456

I guess the easiest solution would be just setting the max memory setting to 0 or some very low value... Could that lead to any problems with other calculations?

**Comment by sunqm** [comment]
Max_memory should only affect the performance. Can you post your test scripts? It
looks like a race-condition bug in the large max_memory tests.

**Comment by HehnLukas** [comment]
Here is an entire code which reproduces the problem (at least for me) The make_natural_orbitals import is just importing the code from "https://github.com/pyscf/pyscf/blob/9a8ebf14724b3d49eeb4a3719d1234c99df6e3e9/pyscf/mcscf/addons.py" for that function, as it is not available in the version installed on my system (1.7.5).  Below I also added the entire bash script for submitting it to our cluster (just changed the paths), really the only difference is the PYSCF_MAX_MEMORY setting:

```
from pyscf import gto, scf
from pyscf.mrpt import nevpt2
from pyscf.scf import uhf
from pyscf import mcscf, lib
from pyscf.tools import mo_mapping
from make_natural_orbitals import make_natural_orbitals
import numpy as np


mol = gto.Mole()
mol.verbose = 4
mol.atom = '''
        O    -3.3006173    2.2577663    0.0000000
        H    -4.0301066    2.8983985    0.0000000
        H    -2.5046061    2.8136052    0.0000000
        '''
mol.build(0,0)
energy = {}

print("SCF Calculation")

# SCF calculation
mf = uhf.UHF(mol)
mf.max_cycle = 500
mf.kernel()

E_SCF = mf.e_tot
    
noons, natorbs = make_natural_orbitals(mf)

homo = 5

# CASSCF
electrons = 6
orbitals = 3
Me_3d_orbitals = mo_mapping.mo_comps('O 2p', mol, natorbs)
Me_4d_orbitals = mo_mapping.mo_comps('O 3p', mol, natorbs)

Me_3d_orbitals = Me_3d_orbitals.argsort()
Me_4d_orbitals = Me_4d_orbitals.argsort()

Me_3d_orbitals = [orb for orb in Me_3d_orbitals if orb <= np.max(homo)]
Me_4d_orbitals = [orb for orb in Me_4d_orbitals if orb > np.max(homo)]

caslist = list(Me_3d_orbitals[-3:])
caslist = [x+1 for x in caslist]

mc1 = mcscf.CASSCF(mf,orbitals,electrons)
mc1.max_cycle = 200
mc1.mo_coeff = natorbs
mc1.mo_occ = noons
mc1.natorb=True
mo = mc1.sort_mo(caslist)
mc1.kernel(mo)
E_CASSCF = mc1.e_tot
print ("CASSCF energy: %f\n" % E_CASSCF)

E_NEVPT2 = nevpt2.NEVPT(mc1).kernel()


log = lib.logger.new_logger(verbose=4)
log.info('current use %d MB', lib.current_memory()[0])

```

```
#!/bin/bash
#PBS -N "HF_pyscf"
#PBS -l mpiprocs=5
#PBS -l ncpus=5
#PBS -l ompthreads=5
#PBS -l mem=25gb
#PBS -l walltime=5:00:00

hostname
module purge
module load rt_gcc_openmpi/7.2.0-3.0.0 pyscf/1.7.5/intelpython3.6-2019.1
export PYSCF_TMPDIR=${PBS_SCRATCH}
export OMP_NUM_THREADS=5
export PYTHONPATH=$(pwd):$PYTHONPATH
cd ${PBS_SCRATCH}

export PYSCF_MAX_MEMORY=0
for i in {1..9}
do
    python ~/path/to/cas_min.py &> ~/path/to/Cas${i}.out
done
export PYSCF_MAX_MEMORY=500
for i in {11..19}
do
    python ~/path/to/cas_min.py &> ~/path/to/Cas${i}.out
done

```

and the outputs are, that the files with max_memory 0 remain constant and with max_memory 500 do not:
Cas11.out:Nevpt2 Energy = -0.025523828600104
Cas12.out:Nevpt2 Energy = -0.031918360725359
Cas13.out:Nevpt2 Energy = -0.031344384814906
Cas14.out:Nevpt2 Energy = -0.031349643884941
Cas15.out:Nevpt2 Energy = -0.031935104625613
Cas16.out:Nevpt2 Energy = -0.031814151950285
Cas17.out:Nevpt2 Energy = -0.031392460989392
Cas18.out:Nevpt2 Energy = -0.031467290296894
Cas19.out:Nevpt2 Energy = -0.031889453178557
Cas1.out:Nevpt2 Energy = -0.031731838110518
Cas2.out:Nevpt2 Energy = -0.031731838110518
Cas3.out:Nevpt2 Energy = -0.031731838110518
Cas4.out:Nevpt2 Energy = -0.031731838110518
Cas5.out:Nevpt2 Energy = -0.031731838110518
Cas6.out:Nevpt2 Energy = -0.031731838110518
Cas7.out:Nevpt2 Energy = -0.031731838110518
Cas8.out:Nevpt2 Energy = -0.031731838110518
Cas9.out:Nevpt2 Energy = -0.031731838110518

The final lines of my code give the memory used at the end, it is not constant in either case, but for both it is always around 90 MB

**Comment by sunqm** [comment]
The entire system. This parameter is only a rough estimator. In many cases, the actual memory use can be more than PYSCF_MAX_MEMORY.

**Comment by emamars95** [comment]
Just a quick comment here. 
I have run some calculation with pbc.SCF and I noticed that the real memory allocated can go more than 150% the PYSCF_MAX_MEMORY that I have set using the max_memory card. Just to give some number I have used 
```
mol, mf_dict = scf.chkfile.load_scf(chk_to_load) # Load the Mol and SCF dictionary from check point file
mol.max_memory = 40000
mol.build()
```
In building the PBC integrals the real memory used is about 100 GB. 
![Capture](https://github.com/pyscf/pyscf/assets/115721827/a9d6868a-ca52-4879-ad9b-8223d4b13493)

I am wondering if that should be a normal behaviour. A quick comment would be helpful for me =)
